core/src/banking_stage/packet_receiver.rs

```diff
@@ -0,0 +1,169 @@
+use {
+    super::BankingStageStats,
+    crate::{
+        immutable_deserialized_packet::ImmutableDeserializedPacket,
+        leader_slot_banking_stage_metrics::LeaderSlotMetricsTracker,
+        packet_deserializer::{BankingPacketReceiver, PacketDeserializer, ReceivePacketResults},
+        tracer_packet_stats::TracerPacketStats,
+        unprocessed_transaction_storage::UnprocessedTransactionStorage,
+    },
+    crossbeam_channel::RecvTimeoutError,
+    solana_measure::{measure, measure::Measure},
+    solana_sdk::{
+        saturating_add_assign,
+        timing::{duration_as_ms, timestamp},
+    },
+    std::{
+        sync::atomic::Ordering,
+        time::{Duration, Instant},
+    },
+};
+
+pub struct PacketReceiver {
+    id: u32,
+    packet_deserializer: PacketDeserializer,
+    last_receive_time: Instant,
+}
+
+impl PacketReceiver {
+    pub fn new(id: u32, receiver: BankingPacketReceiver) -> Self {
+        Self {
+            id,
+            packet_deserializer: PacketDeserializer::new(receiver),
+            last_receive_time: Instant::now(),
+        }
+    }
+
+    pub fn do_packet_receiving_and_buffering(
+        &mut self,
+        unprocessed_transaction_storage: &mut UnprocessedTransactionStorage,
+        banking_stage_stats: &mut BankingStageStats,
+        tracer_packet_stats: &mut TracerPacketStats,
+        slot_metrics_tracker: &mut LeaderSlotMetricsTracker,
+    ) -> Result<(), RecvTimeoutError> {
+        // Gossip thread will almost always not wait because the transaction storage will most likely not be empty
+        let recv_timeout = if !unprocessed_transaction_storage.is_empty() {
+            // If there are buffered packets, run the equivalent of try_recv to try reading more
+            // packets. This prevents starving BankingStage::consume_buffered_packets due to
+            // buffered_packet_batches containing transactions that exceed the cost model for
+            // the current bank.
+            Duration::from_millis(0)
+        } else {
+            // Default wait time
+            Duration::from_millis(100)
+        };
+
+        let (res, receive_and_buffer_packets_time) = measure!(self.receive_and_buffer_packets(
+            recv_timeout,
+            unprocessed_transaction_storage,
+            banking_stage_stats,
+            tracer_packet_stats,
+            slot_metrics_tracker,
+        ));
+        slot_metrics_tracker
+            .increment_receive_and_buffer_packets_us(receive_and_buffer_packets_time.as_us());
```

Comment:
imo, `do_packet_receiving_and_buffering` and `receive_and_buffer_packets` are too similarly named. how about moving this metrics line to the end of `receive_and_buffer_packets`. and make this return Duration with a name like `fn determine_timeout` and to be called from `fn receive_and_buffer_packets()` and make `determine_timeout()` private and `receive_and_buffer_packets()` `pub` to be called by BankingStage. (I think `receive_and_buffer_packets` is more fit to expose as api due to its straightforward grammar)

Reply:
will include this into follow-up PR(s). I agree the names were too similar.